packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -2896,3 +3025,270 @@ typedef SelectableRegionContextMenuBuilder = Widget Function(
+  SelectionListenerStatus _selectionListenerStatus = SelectionListenerStatus.finalized;
+  SelectionListenerStatus get selectionListenerStatus => _selectionListenerStatus;
+  set selectionListenerStatus(SelectionListenerStatus newStatus) {
+    _selectionListenerStatus = newStatus;
+    notifyListeners();
+  }
+}
+
+/// Notifies its listeners when the selection under a [SelectableRegion] or
+/// [SelectionArea] is finalized or has changed.
+class _SelectableRegionScope extends InheritedNotifier<_SelectableRegionSelectionStatusNotifier> {
+  const _SelectableRegionScope({
+    required _SelectableRegionSelectionStatusNotifier selectionStatusNotifier,
+    required super.child,
+  }) : super(notifier: selectionStatusNotifier);
+
+  /// The closest instance of this class that encloses the given context.
+  ///
+  /// If there is no enclosing [SelectableRegion] or [SelectionArea] widget, then null is
+  /// returned.
+  ///
+  /// Calling this method will create a dependency on the closest
+  /// [_SelectableRegionScope] in the [context], if there is one.
+  static _SelectableRegionSelectionStatusNotifier? maybeOf(BuildContext context) {
+    return context.dependOnInheritedWidgetOfExactType<_SelectableRegionScope>()?.notifier;
+  }
+
+  @override
+  bool updateShouldNotify(_SelectableRegionScope oldScope) {
+    return notifier != oldScope.notifier;
+  }
+}
+
+/// A [SelectionContainer] that allows the user to listen to selection changes
+/// for the child subtree it wraps under a [SelectionArea] or [SelectableRegion].
+///
+/// The selection updates are provided through the [selectionNotifier], to listen
+/// to these updates attach a listener through [SelectionListenerNotifier.addListener]
+/// or [SelectionListenerNotifier.addStatusListener].
+///
+/// This widget should have an ancestor [SelectionArea] or [SelectableRegion]
+/// to be able to listen to selection changes in this widgets subtree.
+///
+/// This widget does not listen to selection changes of nested [SelectionArea]s
+/// or [SelectableRegion]s in its subtree because those widgets are self-contained
+/// and do not bubble up their selection. To listen to selection changes of a
+/// [SelectionArea] or [SelectableRegion] under this [SelectionListener], add
+/// an additional [SelectionListener] under each one.
+///
+/// {@tool dartpad}
+/// This example shows how to use [SelectionListener] to listen to selection changes
+/// under a [SelectionArea] or [SelectableRegion].
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.1.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how to color the active selection red
+/// under a [SelectionArea] or [SelectableRegion].
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.2.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///   * [SelectableRegion], which provides an overview of the selection system.
+class SelectionListener extends StatefulWidget {
```

Comment:
Is it worth splitting this out to its own file at this point?

Reply:
I'll consider this when this is close to merging since i'm not sure if it will use private SelectableRegion APIs or not.

Reply:
Last chance to split this out, but I'm fine either way.